File: heaps/remove_stones_to_minimize_total.java
public int minStoneSum(int[] piles, int k)
{
    PriorityQueue<Integer> heap =
        new PriorityQueue<>(Comparator.reverseOrder());
    
    int totalSum = 0;
    for (int num : piles)
    {
        totalSum += num;
        heap.add(num);
    }

    for (int i = 0; i < k; i++)
    {
        int curr = heap.remove();
        int remove = curr / 2;
        total -= remove;
        heap.add(curr - remove);
    }

    return total;
}

File: heaps/remove_stones_to_minimize_total.py
from heapq import *

def minStoneSum(piles: list[int], k: int) -> int:
    heap = [-num for num in piles]
    heapify(heap)

    for _ in range(k):
        curr = -heappop(heap)
        remove = curr // 2
        heappush(heap, -(curr - remove))
    
    return -sum(heap)